src/Refinery/Tactic.hs

```diff
@@ -125,18 +135,44 @@ ensure p f t = check >> t
           s <- get
           modify f
           case p s of
-            Just err -> throwError err
+            Just err -> unsolvable err
             Nothing -> pure e
 
 -- | Apply the first tactic, and then apply the second tactic focused on the @n@th subgoal.
 focus :: (Functor m) => TacticT jdg ext err s m () -> Int -> TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
 focus t n t' = t <@> (replicate n (pure ()) ++ [t'] ++ repeat (pure ()))
 
+-- | @tweak f t@ lets us modify the extract produced by the tactic @t@.
+tweak :: (Functor m) => (ext -> ext) -> TacticT jdg ext err s m () -> TacticT jdg ext err s m ()
+tweak f t = tactic $ \j -> mapExtract' f $ proofState t j
+
+-- | @peek t k@ lets us examine the extract produced by @t@, and then run a tactic based off it's value.
+peek :: (Functor m) => TacticT jdg ext err s m a -> (ext -> TacticT jdg ext err s m b) -> TacticT jdg ext err s m a
+peek t k = (tactic $ \j -> Subgoal ((), j) (\e -> fmap (first (const ())) $ proofState (k e) j)) >> t
+
+-- | @poke t k@ lets us examine the extract produced by @t@, and then run a tactic that produces a new extract.
+poke :: (Functor m) => TacticT jdg ext err s m () -> (ext -> TacticT jdg ext err s m ext) -> TacticT jdg ext err s m ()
```

Comment:
you absolute madman. you did it!

Reply:
Immediately after the stream finished I realized how it would work lol